File: app/src/main/java/org/hugoandrade/rtpplaydownloader/common/ExpandCollapseAnimation.java
package org.hugoandrade.rtpplaydownloader.common;

import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.Transformation;

public class ExpandCollapseAnimation extends Animation {

    private static final String TAG = ExpandCollapseAnimation.class.getSimpleName();

    // The views to be animated
    private View view;

    // The dimensions and animation values (this is stored so other changes to layout don't interfere)
    private final int fromDimension; // Dimension to animate from
    private int toDimension; // Dimension to animate to
    private int finalToDimension; // Dimension to animate to


    // Constructor
    public ExpandCollapseAnimation(View view, int toDimension) {
        this(view, view.getMeasuredHeight(), toDimension);

    }

    // Constructor
    public ExpandCollapseAnimation(View view, int fromDimension, int toDimension) {
        // Setup references
        // the view to animate
        this.view = view;
        // Get the current starting point of the animation (the current parentWidth or height of the provided view)
        this.fromDimension = fromDimension;

        // Dimension to animate to
        if (toDimension == ViewGroup.LayoutParams.WRAP_CONTENT) {
            //int initialHeight = view.getMeasuredHeight();
            view.measure(ViewGroup.LayoutParams.MATCH_PARENT, toDimension);
            this.toDimension = view.getMeasuredHeight();
            this.finalToDimension = toDimension;
        }
        else {
            this.toDimension = toDimension;
            this.finalToDimension = toDimension;
        }
    }

    @Override
    public void applyTransformation(float interpolatedTime, Transformation t) {
        // Used to apply the animation to the view
        // Animate given the height or parentWidth

        // android.util.Log.e(TAG, "applyTransformation = " + (fromDimension + (int) ((toDimension - fromDimension) * interpolatedTime)));
        view.getLayoutParams().height = interpolatedTime == 1?
                finalToDimension :
                fromDimension + (int) ((toDimension - fromDimension) * interpolatedTime);

        //view.setLayoutParams(view.getLayoutParams());

        // Ensure the views are measured appropriately
        view.requestLayout();
    }
}

File: android-app/src/main/kotlin/org/hugoandrade/rtpplaydownloader/widget/ExpandCollapseAnimation.kt
package org.hugoandrade.rtpplaydownloader.widget

import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.Transformation


class ExpandCollapseAnimation// Constructor
(// The views to be animated
        private val view: View, // The dimensions and animation values (this is stored so other changes to layout don't interfere)
        private val fromDimension: Int // Dimension to animate from
        , toDimension: Int) : Animation() {
    private var toDimension: Int = 0 // Dimension to animate to
    private var finalToDimension: Int = 0 // Dimension to animate to

    companion object {

        private val TAG = ExpandCollapseAnimation::class.java.simpleName
    }

    // Constructor
    constructor(view: View, toDimension: Int) : this(view, view.measuredHeight, toDimension) {

    }

    init {

        // Dimension to animate to
        if (toDimension == ViewGroup.LayoutParams.WRAP_CONTENT) {
            //int initialHeight = view.getMeasuredHeight();
            view.measure(ViewGroup.LayoutParams.MATCH_PARENT, toDimension)
            this.toDimension = view.measuredHeight
            this.finalToDimension = toDimension
        } else {
            this.toDimension = toDimension
            this.finalToDimension = toDimension
        }
    }// Setup references
    // the view to animate
    // Get the current starting point of the animation (the current parentWidth or height of the provided view)

    public override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
        // Used to apply the animation to the view
        // Animate given the height or parentWidth

        // android.util.Log.e(TAG, "applyTransformation = " + (fromDimension + (int) ((toDimension - fromDimension) * interpolatedTime)));
        view.layoutParams.height = if (interpolatedTime == 1f)
            finalToDimension
        else
            fromDimension + ((toDimension - fromDimension) * interpolatedTime).toInt()

        //view.setLayoutParams(view.getLayoutParams());

        // Ensure the views are measured appropriately
        view.requestLayout()
    }
}